src/backend/distributed/utils/colocation_utils.c

```diff
@@ -14,14 +14,322 @@
+	forboth(leftShardIntervalCell, leftShardIntervalList,
+			rightShardIntervalCell, rightShardIntervalList)
+	{
+		ShardInterval *leftInterval = (ShardInterval *) lfirst(leftShardIntervalCell);
+		ShardInterval *rightInterval = (ShardInterval *) lfirst(rightShardIntervalCell);
+
+		List *leftPlacementList = NIL;
+		List *rightPlacementList = NIL;
+		List *sortedLeftPlacementList = NIL;
+		List *sortedRightPlacementList = NIL;
+		ListCell *leftPlacementCell = NULL;
+		ListCell *rightPlacementCell = NULL;
+
+		uint64 leftShardId = leftInterval->shardId;
+		uint64 rightShardId = rightInterval->shardId;
+
+		bool shardsIntervalsEqual = ShardsIntervalsEqual(leftInterval, rightInterval);
+		if (!shardsIntervalsEqual)
+		{
+			ereport(ERROR, (errmsg("shard intervals don't match for %s and %s",
+								   leftRelationName, rightRelationName)));
+		}
+
+		leftPlacementList = ShardPlacementList(leftShardId);
+		rightPlacementList = ShardPlacementList(rightShardId);
+
+		/* sort shard placements according to the node */
+		sortedLeftPlacementList = SortList(leftPlacementList,
+										   CompareShardPlacementsByNode);
+		sortedRightPlacementList = SortList(rightPlacementList,
+											CompareShardPlacementsByNode);
+
+		/* compare shard placements one by one */
+		forboth(leftPlacementCell, sortedLeftPlacementList,
+				rightPlacementCell, sortedRightPlacementList)
+		{
+			ShardPlacement *leftPlacement =
+				(ShardPlacement *) lfirst(leftPlacementCell);
+			ShardPlacement *rightPlacement =
+				(ShardPlacement *) lfirst(rightPlacementCell);
+
+			/*
+			 * If shard placements are on different nodes, these shard
+			 * placements are not colocated.
+			 */
+			int nodeCompare = CompareShardPlacementsByNode((void *) &leftPlacement,
+														   (void *) &rightPlacement);
+			if (nodeCompare != 0)
+			{
+				ereport(ERROR, (errmsg("some shard placements are not colocated for %s "
+									   "and %s", leftRelationName, rightRelationName)));
+			}
+
+			/* we also don't allow colocated shards to be in different shard states */
+			if (leftPlacement->shardState != rightPlacement->shardState)
+			{
+				ereport(ERROR, (errmsg("%s and %s have shard placements in "
+									   "different shard states",
+									   leftRelationName, rightRelationName)));
+			}
+		}
+	}
+}
+
+
+/*
+ * ShardsIntervalsEqual checks if two shard intervals of hash distributed
+ * tables are equal.
+ *
+ * We do min/max value check here to decide whether two shards are colocated,
+ * instead we can simply use FindShardIntervalIndex function on both shards then
+ * but do index check, but we avoid it because this way it is more cheaper.
+ *
+ * Having colocated tables implies that tables are partitioned by hash partition
+ * therefore it is safe to use DatumGetInt32 here.
+ */
+static bool
+ShardsIntervalsEqual(ShardInterval *leftShardInterval, ShardInterval *rightShardInterval)
+{
+	int32 leftShardMinValue = DatumGetInt32(leftShardInterval->minValue);
```

Comment:
I'm skipping this after out conversation. The paths to this function check for hash distribution. We documented that it expects hash distributed tables for future uses. Also keeping this in this file because it is only used locally (at least for the time being).